### DISCUSSION TITLE
Proposal of a mapping between collection(List) and POJO

### DISCUSSION THREAD

**Discussion by meilleureVie** [mapstruct/mapstruct#3318] [Ideas]
Sometimes we assume that in a list we only have one item and we want to map that item to a pojo.
Let take an example of a parcel with its owner.
A parcel can have multiple ancient owners but only one current owner.
```java
public class Parcel { 
    public Long id; 
    public String parcelInfo; 
    public List<User> owners; 
}
public class  User { 
    public Long id; 
    public String name; 
    public Boolean currentOwner;
}
```
Now suppose that in the list of owners, we only have one owner since we are only retrieving the current owner.
Then let map our parcel to the following ParcelDto represented like this :
```java
public class ParcelDto {
    public Long id; 
    public String parcelInfo; 
    public String ownerName; 
}
```
Right now in order to map Parcel to ParcelDto,  we have to do something like this in the parcel mapper :
```java
@Mapping(target = "ownerName", expression = "java(ownerNameFromListWithJustOneItem(parcel.owners))")
public abstract ParcelDto toParcelDto(Parcel parcel);

public String ownerNameFromListWithJustOneItem(List<User> users) {
    assert users.size() == 1;
    return users.get(0).name;
}
```
That is good but i think you can improve the feature to do more and make us save code.
Why not accept a syntax like this that can help us to achieve the same goal in one line ?
```java
@Mapping(target = "ownerName", source ="owners[0].name")
public abstract ParcelDto toParcelDto(Parcel parcel);
```
or this
```java
@Mapping(target = "ownerName", source ="owners^[].name")
public abstract ParcelDto toParcelDto(Parcel parcel);
```
and even add some predifine filter expression query
```java
@Mapping(target = "ownerName", source ="owners^[currentOwner == true].name")
public abstract ParcelDto toParcelDto(Parcel parcel);
```
Caption
[0] or ^[] stands for first item of the list
[n] or $[] stands for last item of the list
[n-1] : before last item and so on

**Comment by filiphr** [comment]
I think that we have an issue for requesting something like that, but I couldn't find it. I think that the idea of this is OK. Your examples are a bit more complex, the only thing that I think makes sense to support is the `[<index>]` syntax, also supporting negative indices. With negative indices the last item would be `[-1], basically we would do `collection.get(collection.size() - <index>)`. 

Things like expressions and filtering are a step too far in my opinion
